project_template/inc/az/exampleshortname/operations.h

```diff
@@ -0,0 +1,20 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/** @file operations.h */
+#pragma once
+#include <stdint.h>

```

Comment:
Consider having #include <cstdint> when a C++ interface is provided

Reply:
sorry I don't understand?

Perhaps this comment got pinned to the wrong line on github.

Reply:
Perhaps is because you should include `stdint.h` for C and `cstdint` for C++.

Reply:
Well in this case we want `<stdint.h>` in C++ as well. <cstdint> is not guaranteed to define ::uint32_t, only std::uint32_t.

Reply:
The lastest C++standard version that I have (C++14) states in "18.4.1 Header <cstdint> synopsis" the following:

"The header defines all functions, types, and macros the same as 7.18 in the C standard."

So based on the above I would reasonably expect that ::uint32_t/uint32_t (in C++-ese) type exists.

---
In reply to: [309482553](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r309482553) [](ancestors = 309482553)

Reply:
fun fact: C++98, C++03 are both mum about cstdint. It appears in C++11. So I guess it all depends on which version of C++ is available/targeted.

---
In reply to: [310337767](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r310337767) [](ancestors = 310337767,309482553)

Reply:
From what I see Charlie is right. Unfortunately including cstdint is not enough, one would have to have a using namespace std to have the same code compile.
So just including stdint.h in C++ too is actually a better option I think.

It would have been nice that the C++ standard would state that cstdint should define ::int32_t etc. Because these being in std forces one of the 3:
- Using stdint.h which defeats the purpose of cstdint
- Adding std:: everywhere for C++, which defeats again the purpose of cstdint
- Have using namespace std all over the place which can introduce other potential collisions

Sooooo, yeah. The lesser evil is #include <stdint.h> I guess. Now we have some code bases to update ...

---
In reply to: [310337826](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r310337826) [](ancestors = 310337826,310337767,309482553)

Reply:
yep. <cmath> is somewhat of a special case since it defines additional overloads. This is also unfortunate but if you're c++ code and want to use those overloads then you need cmath, and not math.h.

fwiw if you're a C library, providing a C interface, it's reasonable to assume your compiler supports C (and its standard library).

Reply:
C++ (at least 14) say stdint.h is provided (and the others). Annex D, D.1 which is normative.
So it should be safe to include stdint.h all the time (in C and C++). 

It is unfortunate that cstdint does not help in this case for C++ as that would be ideal. There are cases where a translation unit must be compiled with the C++ compiler and still needs to include headers from the C lib. Happened several times when porting IoT SDK for C that such interactions were needed. From a nicety and consistency point of view would have been nice to be able to use cstdint in those cases, but ... it is what it is.

---
In reply to: [310861226](https://github.com/Azure/azure-sdk-for-c/pull/2#discussion_r310861226) [](ancestors = 310861226)